torch/csrc/jit/runtime/profiling_record.cpp

```diff
@@ -146,6 +164,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, Value* i) {
   n->replaceInputWith(i, pn->output());
 }
 
+static void removeProfileNodes(Block* b) {
+  for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+    if (it->kind() == prim::profile) {
+      if (it->outputs().size() == 1) {
+        it->output()->replaceAllUsesWith(it->input(0));
+      }
+      it.destroyCurrent();
+    } else {
+      for (auto ib : it->blocks()) {
+        removeProfileNodes(ib);
+      }
+    }
+  }
+}
+
+void ProfilingRecord::peelLoops() {
+  // we peel starting out from the outermost loops, so we can avoid having
+  // peeled loops in outer loop peels
+  for (auto it = loops_with_profiled_nodes_.rbegin();
+       it != loops_with_profiled_nodes_.rend();
+       ++it) {
+    auto peel = PeelLoop(*it, 1);
+    // N.B. the reason why we need to remove `prim::profile`
```

Comment:
I am not a fan of the coupling of loop peeling with profile information. There are two ways this harms it:

1. when to loop peel and  when profile are bound together, but in the future there is not reason this may always be the case, we may develop these heuristics separately. For instance, we may profile any node going into an if-statement.
2. This code has to fix up profiling information of peeled loops, which is tricky to understand.

My suggestion would be to do this in a different order. Before inserting profiling nodes, peel the loops.  It should be pretty easy to just decide to peel loops using the same criteria used to insert prim::profile nodes. Generally it makes sense to peel any look with tensor compute in it, so we can look for that. Then insert prim::profile nodes. This avoids both of the couplings above, and allow someone to adu

Reply:
factored out this logic into the `ProfilingLoopPeeler` class